File: 39-combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        
        result = []

        def comb(index, current, total):
            if total == target:
                result.append(current[:])
                return
            if index >= len(candidates) or total > target:
                return 

            current.append(candidates[index])
            comb(index, current, total + candidates[index])
            current.pop()
            comb(index + 1, current, total)

        comb(0, [], 0)
        return result

File: 39-combination-sum/combination-sum.java
class Solution {

    public void backtrack(List<Integer> current, int[] candidates, int target, List<List<Integer>> result, int index) {

        if (target == 0) {
            result.add(new ArrayList<>(current));
            return;
        }
        if (index >= candidates.length || target < 0) {
            return;
        }

        current.add(candidates[index]);
        target -= candidates[index];
        backtrack(current, candidates, target, result, index);
        current.remove(current.size() - 1);
        target += candidates[index];
        backtrack(current, candidates, target, result, index + 1);
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {

        List<List<Integer>> result = new ArrayList<>();
        backtrack(new ArrayList<>(), candidates, target, result, 0);
        return result;
    }
}